### PR TITLE
Implement W-MSE Loss and Transform

### PR DESCRIPTION
**Pull request by johnsutor**
This pull request implements the W-MSE Transform and W-MSE Loss as proposed in the paper https://arxiv.org/pdf/2007.06346.pdf. The code largely comes from the original repository https://github.com/htdt/self-supervised, preserving the default hyperparameters with some minor changes to make it compatible with lightly. This feature was recommended in this issue: https://github.com/lightly-ai/lightly/issues/1424#event-11223331529

### CODE REVIEW

**Comment by philippmwirth** [comment]
Hi @johnsutor, thank you very much for your contribution. Someone from our team will have a look asap.

**Review by philippmwirth — CHANGES_REQUESTED** [review]
Thank you for the changes, @johnsutor! The PR looks great so far. I have left a few comments on things I think we should change before merging. If you have any questions or require help updating the PR please let me know 🙂

File: tests/loss/test_WMSELoss.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        bs = 128
+        dim = 64
+        num_samples = 32
+
+        loss_fn = WMSELoss()
+        x = torch.randn(bs * num_samples, dim)
+
+        loss = loss_fn(x)
+
+        print(loss)
```
[the remark this answers is not in the document]
Comment (johnsutor) [reply tests/loss/test_WMSELoss.py:19]: Sorry I left those!

File: lightly/loss/wmse_loss.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+
+class Whitening2d(nn.Module):
+    """
+    Implementation of the whitening layer as described in [0].
+
+    [0] Whitening for Self-Supervised Representation Learning, 2021, https://arxiv.org/pdf/2007.06346.pdf
+    """
+
+    running_mean: torch.Tensor
```
[the remark this answers is not in the document]
Comment (johnsutor) [reply lightly/loss/wmse_loss.py:24]: Precisely, will do.

File: lightly/loss/wmse_loss.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        >>> h = torch.cat([model(s) for s in samples])
+        >>>
+        >>> # calculate loss
+        >>> loss = loss_fn(samples, h)
+
+    """
+
+    def __init__(
+        self,
+        embedding_dim: int = 64,
```
[the remark this answers is not in the document]
Comment (johnsutor) [reply lightly/loss/wmse_loss.py:112]: Just double checked https://arxiv.org/pdf/2007.06346.pdf, and in the experiment implementation details, they list that they set the embedding_dim to be 128 for imagenet. I'll change it to 128 if that's OK.

File: lightly/transforms/wmse_transform.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+                T.RandomApply(
+                    [T.ColorJitter(cj_bright, cj_contrast, cj_sat, cj_hue)], p=cj_prob
+                ),
+                T.RandomGrayscale(p=random_gray_scale),
+                T.RandomResizedCrop(
+                    input_size,
+                    scale=(min_scale, 1.0),
+                    interpolation=3,
+                ),
+                T.RandomHorizontalFlip(p=hf_prob),
```
[the remark this answers is not in the document]
Comment (johnsutor) [reply lightly/transforms/wmse_transform.py:85]: Based on the results from here: https://stackoverflow.com/questions/62968174/for-pil-imagefilter-gaussianblur-how-what-kernel-is-used-and-does-the-radius-par I opted to stick with the default GaussianBlur transform already in lightly.

File: lightly/loss/wmse_loss.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (johnsutor) [reply lightly/loss/wmse_loss.py:112]: Also changing the default w_size based on this line from the paper:

> Next,each partition is further split in sub-batches, using the heuristic that the size of each sub-batch (Vi) should be equal to the size of embedding (|v|) times 2 (this prevents instability issues when computing the covariance matrices).

[the remark this answers is not in the document]
Comment (philippmwirth) [reply lightly/loss/wmse_loss.py]: Thanks!

**Review by philippmwirth — APPROVED** [review]
This looks good now! I'll quickly do a sanity check on Cifar10 and then we're good to merge. Thank you 🙂

**Comment by philippmwirth** [comment]
The results look promising, after 3 epochs we get a kNN accuracy of over 50% with `num_samples=4` and SimCLR projection head.
```
------------------------------------------------------------------------------------------
| Model         | Batch Size | Epochs |  KNN Test Accuracy |       Time | Peak GPU Usage |
------------------------------------------------------------------------------------------
| WMSE          |        512 |      3 |              0.519 |   12.9 Min |      9.4 GByte |
------------------------------------------------------------------------------------------
```
Will merge this now and leave the proper benchmarking for later. @guarin FYI.
